### DISCUSSION TITLE
"Too Many Requests" although there are only 4-5 concurrent requests at max

### DISCUSSION THREAD

**Discussion by rahul-san** [deepgram/community#586] [General help] [answered]
we are using "pay-as-you-go" plan. 

curl \
-X POST \
-H "Authorization: Token <MY KEY>" \
-H 'content-type: application/json' \
-d '{"url":"<presigned aws s3 url>"}' \
"https://api.deepgram.com/v1/listen?language=en-US&smart_format=false&paragraphs=true&model=nova-2&punctuate=true&diarize=true&detect_entities=true"

{"err_code":"Too Many Requests","err_msg":"Please try again later.","request_id":"14554824-2bd2-47ec-98e1-0d0c7206a571"}

I retry the request after 10s and 2 min, still the same err comes.
Please Help 🙂

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by jpvajda** [comment]
@rahul-san  👋  this is a `429 Too Many Requests:` The user has sent too many requests in a given amount of time error. 

It is recommended to implement an exponential-backoff retry strategy when submitting a large volume of concurrent requests to accommodate rate-limiting. An exponential-backoff strategy introduces a delay before each retry attempt. The delay between retries increases exponentially with each consecutive failure, helping to mitigate the impact of transient errors and reduce the likelihood of overwhelming the system with retries.

The model you used in the request is also a consideration:  You can see our [Docs on Model Rate limits]( https://developers.deepgram.com/docs/getting-started-with-pre-recorded-audio#rate-limits)

*  For Nova, Base, and Enhanced, the rate limit is 100 concurrent requests.
*  For Whisper, the rate limit is 15 concurrent requests with a paid plan and 5 concurrent requests with the pay-as-you-go plan.

Let me know if this helps.

**Reply by rahul-san** [reply]
Thanks for replying @jpvajda 
I have already implemented `exponential-backoff strategy`. Usually there are only 4-5 concurrent requests at max. Do Nova-2 model also have max 5 concurrent requests?
I also created new api keys and project and deleted the old one. Still same issue.
Is there a way to check the api logs, which shows rate limited calls also?

**Comment by wesbos** [comment]
@jpvajda  We're receiving this error with only 1 request - anyone else?

**Reply by wesbos** [reply]
seems to be working now, but looking at our Deepgram logs, we only see 1 request, but we got this error ~5 times this morning. 

<img width="351" alt="Screenshot 2024-02-09 at 11 28 05 AM" src="https://github.com/deepgram/community/assets/176013/7716d122-1e1e-4430-88c7-c7cbed68d6c7">

**Comment by jpvajda** [comment, marked as answer]
@rahul-san  I dug into the request you made and it seems you used `entity_detection`.  Our Audio Intelligence features have different rate limits as I believe entity requests are currently set to 5 concurrent requests.

**Comment by jpvajda** [comment]
@wesbos  👋  Could you provide me with a `Request ID` and I can look into it deeper to see what info I can find for you.

**Reply by wesbos** [reply]
DG: {"err_code":"Too Many Requests","err_msg":"Please try again later.","request_id":"ca726f71-93e5-40c7-934a-9685d4dd43a2"}

Here is one we are getting - could you check?

**Comment by jpvajda** [comment]
@wesbos  It looks like you did run `detect_entities=true` in this request and this particular features is rate limited at a different rate, of up to 10 concurrent requests depending on the type of account you have with Deepgram.

Below are the query params used in this request, note I removed all the keywords since this a public post.

```
punctuate=true&model=nova-2-ea&language=en-US&detect_entities=true&diarize=true&smart_format=true&paragraphs=true&utterances=true&detect_topics=false&keywords= ...
```

**Reply by wesbos** [reply]
Thanks - we are firing off the request at most every 10 mins, and it seems this pops up when our serverless function times out (our problem, not yours). But there is no way we could be running them concurrently, even 2 at a time. Seems if I fire of any requests right now I get that issue. Is it possible when our function times out, you aren't "closing" it on your end?

**Comment by DamienDeepgram** [comment]
> Thanks - we are firing off the request at most every 10 mins, and it seems this pops up when our serverless function times out (our problem, not yours). But there is no way we could be running them concurrently, even 2 at a time. Seems if I fire of any requests right now I get that issue. Is it possible when our function times out, you aren't "closing" it on your end?

What API parameters are you using? If you are making async requests with a callback url those are considered concurrent requests until the audio has been processed.

So if you send us audio to process async and we immediately send you a response with a Request ID that audio is still being processed.

You would need to wait until the callback is called with the transcript before considering the async request complete

**Reply by wesbos** [reply]
I will check, but it's using the JS API and awaiting the promise - no callback url used. So it's possible we have rogue requests we can't cancel? Does that make sense?

**Comment by jpvajda** [comment]
@wesbos 👋  @DamienDeepgram  and I did some troubleshooting on our end we aren't seeing anything that might be causing these issues you are seeing. For example we ran this [~60 min podcast ]( https://d11untcg2uthr3.cloudfront.net/attachments/41468731/no-such-thing-as-a-human-sized-peanut.mp3?session_id=7c70ba3c-5d21-5674-a829-97eae6575922&ab=128&al=57027295&ao=73123&ct=DOWNLOAD&episode_id=8469281&show_id=2399216&user_id=uk_nosuchthingasafish&organization_id=audioboom&tenant=AUDIOBOOM&timestamp=1710283992&metadata=dist%3Dsponsored_ads%26one_min%3D1033123%26pl%3Drss%26sub%3Dex&response-content-disposition=inline&media_type=static&Expires=1710456792&Key-Pair-Id=K1J2BR3INU6RYD&Signature=Y7SX9mShQUy5mfvInD7Awq8w~pddODKCjv3L~7T9bwYQX9tofvkt4QFBDhrQz1kYuP4zHrm~ZWhozcjQbKfTMKzEcJ4jBzphQIDniUAdJ~nSql6WrNie8p3bsU9B2E6bedC4IGZUyyH-ufY0VnWqTrxL0-yoGdGkSxci~Ajg9Ln~TIWSJ2gxaC6dxQ5P6Yb5CmXY7IYjqBSZ7ULECqslCNWMxnt8p5QTkRJi5vUrBkfg93YyH4NEs1vng34PfAcbDdE~JJ~EeAPOVD4obhx0e44~FkBkcp7XFlUUe3pLF4Q-6RTdECnnYnG5opg60vXC6eZkcEhprHn5eG1lh7hINw__)  With the same API parameters you used and it processed in `2m 24s` You might want to try collecting log data and request IDs to see if you can pinpoint the issue. 

If you can provide any additional information to us that could point to an issue with the Deepgram Service feel free to send that to `devrel@deepgram.com` if you don't want to post that in a public Github Discussions post.

**Reply by wesbos** [reply]
thanks - the issue is that our serverless function times out before deepgram comes back - which is 100% our issue and I can fix that - but because they kept timing out, I'm now in a situation where no single request to deepgram will go though even with enough server time. 

Just hit another one: DG: {"err_code":"Too Many Requests","err_msg":"Please try again later.","request_id":"eea4ecb9-f5ef-4dd8-b6df-5f51a4d4d1a1"}

**Reply by wesbos** [reply]
Here is another: DG: {"err_code":"Too Many Requests","err_msg":"Please try again later.","request_id":"8ea824b7-57d3-485a-9bac-0f2a3d8c233a"} 

This one happned about 4 mins before our last error. I cannot see 5 of them happening at once, let alone in the same 20 min span

**Reply by wesbos** [reply]
Here is some more logging:

We request the transcript every 20 mins:

<img width="1614" alt="Screenshot 2024-03-12 at 8 51 16 PM" src="https://github.com/deepgram/community/assets/176013/a4463b37-b4ec-420d-89ba-df2e5dde18df">

And every single one returns this error:


<img width="2081" alt="Screenshot 2024-03-12 at 8 53 03 PM" src="https://github.com/deepgram/community/assets/176013/f77a7da0-3313-4791-95df-de2c9916198f">


DG: {"err_code":"Too Many Requests","err_msg":"Please try again later.","request_id":"8ea824b7-57d3-485a-9bac-0f2a3d8c233a"}
DG: {"err_code":"Too Many Requests","err_msg":"Please try again later.","request_id":"e4a485a9-fa14-412f-bee1-c6928d342623"}
DG: {"err_code":"Too Many Requests","err_msg":"Please try again later.","request_id":"ef653892-12ce-4a06-97e0-f899190c8af8"}
DG: {"err_code":"Too Many Requests","err_msg":"Please try again later.","request_id":"deb6cf74-b353-48f1-b498-533769a50630"}
DG: {"err_code":"Too Many Requests","err_msg":"Please try again later.","request_id":"1a800787-44ae-43dd-98f1-1f8b8459df64"}
DG: {"err_code":"Too Many Requests","err_msg":"Please try again later.","request_id":"1a4f9a30-9a04-4efa-a53d-7ceb091123ca"}
DG: {"err_code":"Too Many Requests","err_msg":"Please try again later.","request_id":"9b7753b9-5287-4bb5-8201-73d9aa87a841"}
DG: {"err_code":"Too Many Requests","err_msg":"Please try again later.","request_id":"d91c6f67-53be-45e3-9e39-83f42a4102d7"}
DG: {"err_code":"Too Many Requests","err_msg":"Please try again later.","request_id":"e0f1f806-2e36-4d46-904f-24dd0de6dcd9"}
DG: {"err_code":"Too Many Requests","err_msg":"Please try again later.","request_id":"4310b8f7-ab52-4763-98ee-fb04654c6a05"}

**Reply by wesbos** [reply]
To add to this, the dashboard only shows that we have made 4 requests in total. So if the limit is 5 concurrent, how are we hitting this issue with 4 requests in a single day? 
<img width="901" alt="Screenshot 2024-03-12 at 9 04 13 PM" src="https://github.com/deepgram/community/assets/176013/cbbc4e98-54bb-4e33-81cc-4c03eb9a06df">

**Comment by jpvajda** [comment]
@wesbos  thanks for all the additional details here, Let me see if I can pull in another one of our Applied Engineers to evaluate what might be occurring here.  @jjmaldonis  / @jkroll-deepgram

**Comment by jkroll-deepgram** [comment]
Hi @wesbos, is this issue still affecting you?

I can see that you are indeed getting 429s on our side. The only potential culprit I can see is the `detect_entities=true`. If you make subsequent requests with this parameter removed, do they succeed?

It is possible that Deepgram has continued to process your request after your serverless function times out. However, I do agree it's odd that the Console dashboard is showing only 4 requests for your usage. The issue you're observing may be partly due to our rate limits, and partly due to how we are handling requests after losing the connection from your serverless function.

**Reply by wesbos** [reply]
Hey -thanks for the help. It seems to have gone through now but I'm pretty sure it has something to do with our serverless function timing out and then causing deepgram to think I am still connected. 

Its happened twice over the last few months - both times because our function timed out.
